File: stonequest/src/com/barelyconscious/game/spawnable/Entity.java
/* *****************************************************************************
 * Project:          Roguelike2.0
 * File name:        Entity.java
 * Author:           Matt Schwartz
 * Date created:     07.19.2012
 * Redistribution:   You are free to use, reuse, and edit any of the text in
                     this file.  You are not allowed to take credit for code
                     that was not written fully by yourself, or to remove 
                     credit from code that was not written fully by yourself.  
                     Please email schwamat@gmail.com for issues or concerns.
 * File description: 
 **************************************************************************** */

package com.barelyconscious.game.spawnable;

import com.barelyconscious.game.Game;

public class Entity extends Sprite {
    private int level;
    private float healthPoints;
    private int reach = 1;
    
    public Entity(String objectName, int tileId) {
        super(objectName, tileId);
        super.setCollision(true);
    } // constructor
    
    public Entity(String objectName, int tileId, int x, int y) {
        super(objectName, tileId);
        super.setCollision(true);
        super.setPosition(x, y);
    } // constructor
    
    public void changeHealthBy(float amount) {
        healthPoints += amount;
        if (healthPoints < 0) {
            healthPoints = 0;
            remove();
        } // if
    } // changeHealthBy
    
    public int getLevel() {
        return level;
    } // getLevel
    
    public void setLevel(int level) {
        this.level = level;
    } // setLevel
    
    /* Remove this entity from the 
        World Map */
    @Override
    public void tick() {
        if (healthPoints <= 0) {
            remove();
        } // if
    } // checkCanRemove
    
    /**
     * Used by subclasses when creating the entity.
     * @param health 
     */
    protected void setHealthPoints(float health) {
        healthPoints = health;
    } // setHealthPoints
    
    protected float getHealthPoints() {
        return healthPoints;
    } // getHealthPoints
    
    public void moveUp() {
        changePositionBy(0, -1);
    } // moveUp
    
    public void moveDown() {
        changePositionBy(0, 1);
    } // moveDown
    
    public void moveLeft() {
        changePositionBy(-1, 0);
    } // moveLeft
    
    public void moveRight() {
        changePositionBy(1, 0);
    } // moveRight
    
    /**
     * How far the entity can be and still reach the target.  Minimum value is 1
     * which is also the default value.
     */
    public void setReach(int r) {
        reach = r;
    } // setReach
    
    /**
     * How far an Entity can perform action() from another Entity
     * @return 
     */
    public int getReach() {
        return reach;
    } // getReach
    
    @Override
    public void remove() {
//        Sound.ENTITY_DEATH.play();
        Game.world.removeEntity(this);
    } // onDeath

    /**
     * This function should probably be decommissioned.
     * @return 
     */
    @Override
    public String toString() {
        return super.getName() + "@coord: [" + super.getXPos() + ", " + 
                super.getYPos() + "]";
    } // toString
} // Entity

File: StoneQuest/src/com/barelyconscious/game/spawnable/Entity.kt
package com.barelyconscious.game.spawnable

import com.barelyconscious.game.Game
import com.barelyconscious.services.SoundService
import com.barelyconscious.systems.audio.PlayableSound
import com.barelyconscious.systems.messaging.MessageSystem
import com.barelyconscious.systems.messaging.data.SoundMessageData

/**
 * Creates a new Entity with the following parameters
 * @param displayName the displayName of the Entity
 * @param tileId the id of the artwork for the Entity
 */
abstract class Entity(
    displayName: String,
    tileId: Int,
    protected val messageSystem: MessageSystem
) : Sprite(displayName, tileId) {

    var level: Int = 0
        protected set
    var lightRadius: Int = 10
    open var maxHealth: Double = 0.0
        protected set
    open var currentHealth: Double = 0.0
        protected set

    private val faction: Int = 0

    init {
        /* All Entities cause collision with the Player and each other */
        super.setCollision(true)
    }

    constructor(
        displayName: String,
        tileId: Int,
        x: Int,
        y: Int,
        messageSystem: MessageSystem
    ) : this(
        displayName,
        tileId,
        messageSystem
    ) {
        super.setCollision(true)
        super.setPosition(x, y)
    }

    /**
     * Change the health of the Entity by amount
     * @param amount the amount to change the health of the Entity by
     */
    open fun changeHealthBy(amount: Double) {
        currentHealth += amount
    }

    /**
     * Whatever the Entity needs to do on a game tick, including
     * marking the Entity for removal if its health points fall below 0.
     */
    override fun tick() {
        if (currentHealth <= 0) {
            remove()
        }
    }

    /**
     * Removes the Entity from the world.
     */
    override fun remove() {
        messageSystem.sendMessage(
            SoundService.PLAY_SOUND,
            SoundMessageData(PlayableSound.ENTITY_DEATH),
            this)
        Game.world.removeEntity(this)
    }
}
